File: uniffi_bindgen/src/bindings/python/templates/RustBufferStream.py

class RustBufferStream(object):
    # Helper for structured reading of bytes from a RustBuffer

    def __init__(self, rbuf):
        self.rbuf = rbuf
        self.offset = 0

    def remaining(self):
        return self.rbuf.len - self.offset

    def _unpack_from(self, size, format):
        if self.offset + size > self.rbuf.len:
            raise InternalError("read past end of rust buffer")
        value = struct.unpack(format, self.rbuf.data[self.offset:self.offset+size])[0]
        self.offset += size
        return value

    def read(self, size):
        if self.offset + size > self.rbuf.len:
            raise InternalError("read past end of rust buffer")
        data = self.rbuf.data[self.offset:self.offset+size]
        self.offset += size
        return data

class RustBufferTypeReader(object):
    # For every type used in the interface, we provide helper methods for conveniently
    # reading that type in a buffer. Putting them on this internal helper object (rather
    # than, say, as methods on the public classes) makes it easier for us to hide these
    # implementation details from consumers, in the face of python's free-for-all type
    # system.
    # This class holds the logic for *how* to read the types from a buffer - the buffer itself is
    # always passed in, because the actual buffer might be owned by a different crate/module.

    {%- for typ in ci.iter_types() -%}
    {%- let canonical_type_name = typ.canonical_name()|class_name_py -%}
    {%- match typ -%}

    {% when Type::Int8 -%}

    @staticmethod
    def readI8(stream):
        return stream._unpack_from(1, ">b")

    {% when Type::UInt8 -%}

    @staticmethod
    def readU8(stream):
        return stream._unpack_from(1, ">B")

    {% when Type::Int16 -%}

    @staticmethod
    def readI16(stream):
        return stream._unpack_from(2, ">h")

    {% when Type::UInt16 -%}

    @staticmethod
    def readU16(stream):
        return stream._unpack_from(2, ">H")

    {% when Type::Int32 -%}

    @staticmethod
    def readI32(stream):
        return stream._unpack_from(4, ">i")

    {% when Type::UInt32 -%}

    @staticmethod
    def readU32(stream):
        return stream._unpack_from(4, ">I")

    {% when Type::Int64 -%}

    @staticmethod
    def readI64(stream):
        return stream._unpack_from(8, ">q")

    {% when Type::UInt64 -%}

    @staticmethod
    def readU64(stream):
        return stream._unpack_from(8, ">Q")

    {% when Type::Float32 -%}

    @staticmethod
    def readF32(stream):
        v = stream._unpack_from(4, ">f")
        return v

    {% when Type::Float64 -%}

    @staticmethod
    def readF64(stream):
        return stream._unpack_from(8, ">d")

    {% when Type::Boolean -%}

    @staticmethod
    def readBool(stream):
        v = stream._unpack_from(1, ">b")
        if v == 0:
            return False
        if v == 1:
            return True
        raise InternalError("Unexpected byte for Boolean type")

    {% when Type::String -%}

    @staticmethod
    def readString(stream):
        size = stream._unpack_from(4, ">i")
        if size < 0:
            raise InternalError("Unexpected negative string length")
        utf8Bytes = stream.read(size)
        return utf8Bytes.decode("utf-8")


    {% when Type::Timestamp -%}
    # The Timestamp type.
    # There is a loss of precision when converting from Rust timestamps
    # which are accurate to the nanosecond,to Python datetimes which have
    # a variable precision due to the use of float as representation.

    @staticmethod
    def read{{ canonical_type_name }}(stream):
        seconds = stream._unpack_from(8, ">q")
        microseconds = stream._unpack_from(4, ">I") / 1000
        # Use fromtimestamp(0) then add the seconds using a timedelta.  This
        # ensures that we get OverflowError rather than ValueError when
        # seconds is too large.
        if seconds >= 0:
            return datetime.datetime.fromtimestamp(0, tz=datetime.timezone.utc) + datetime.timedelta(seconds=seconds, microseconds=microseconds)
        else:
            return datetime.datetime.fromtimestamp(0, tz=datetime.timezone.utc) - datetime.timedelta(seconds=-seconds, microseconds=microseconds)

    {% when Type::Duration -%}
    # The Duration type.
    # There is a loss of precision when converting from Rust durations
    # which are accurate to the nanosecond,to Python durations which have
    # are only accurate to the microsecond.

    @staticmethod
    def read{{ canonical_type_name }}(stream):
        return datetime.timedelta(seconds=stream._unpack_from(8, ">Q"), microseconds=(stream._unpack_from(4, ">I") / 1.0e3))

    {% when Type::Object with (object_name) -%}
    # The Object type {{ object_name }}.

    @staticmethod
    def read{{ canonical_type_name }}(stream):
        # The Rust code always expects pointers written as 8 bytes,
        # and will fail to compile if they don't fit in that size.
        pointer = stream._unpack_from(8, ">Q")
        return {{ object_name|class_name_py }}._make_instance_(pointer)

    {% when Type::Enum with (enum_name) -%}
    {%- let e = ci.get_enum_definition(enum_name).unwrap() -%}
    # The Enum type {{ enum_name }}.

    @staticmethod
    def read{{ canonical_type_name }}(stream):
        variant = stream._unpack_from(4, ">i")
        {% if e.is_flat() -%}
        return {{ enum_name|class_name_py }}(variant)
        {%- else -%}
        {%- for variant in e.variants() %}
        if variant == {{ loop.index }}:
            {%- if variant.has_fields() %}
            return {{ enum_name|class_name_py }}.{{ variant.name()|enum_name_py }}(
                {%- for field in variant.fields() %}
                stream.read{{ field.type_().canonical_name()|class_name_py }}(){% if loop.last %}{% else %},{% endif %}
                {%- endfor %}
            )
            {%- else %}
            return {{ enum_name|class_name_py }}.{{ variant.name()|enum_name_py }}()
            {% endif %}
        {%- endfor %}
        raise InternalError("Unexpected variant tag for {{ canonical_type_name }}")
        {%- endif %}

    {% when Type::Error with (error_name) -%}
    {%- let e = ci.get_error_definition(error_name).unwrap().wrapped_enum() %}

    # The Error type {{ error_name }}

    # Top-level read method
    @classmethod
    def read{{ canonical_type_name }}(cls, stream):
        variant = stream._unpack_from(4, ">i")
        try:
            read_variant_method = getattr(cls, 'readVariant{}Of{{canonical_type_name}}'.format(variant))
        except AttributeError:
            raise InternalError("Unexpected variant value for error {{ canonical_type_name }} ({})".format(variant))
        return read_variant_method(stream)

    # Read methods for each individual variants
    {%- for variant in e.variants() %}

    @classmethod
    def readVariant{{ loop.index}}Of{{ canonical_type_name }}(cls, stream):
        {%- if variant.has_fields() %}
        return {{ error_name|class_name_py }}.{{ variant.name()|class_name_py }}(
            {%- for field in variant.fields() %}
            cls.read{{ field.type_().canonical_name()|class_name_py }}(stream),
            {%- endfor %}
        )
        {%- else %}
        return {{ error_name|class_name_py }}.{{ variant.name()|class_name_py }}()
        {%- endif %}
    {%- endfor %}

    {% when Type::Record with (record_name) -%}
    {%- let rec = ci.get_record_definition(record_name).unwrap() -%}
    # The Record type {{ record_name }}.

    @classmethod
    def read{{ canonical_type_name }}(cls, stream):
        return {{ rec.name()|class_name_py }}(
            {%- for field in rec.fields() %}
            cls.read{{ field.type_().canonical_name()|class_name_py }}(stream){% if loop.last %}{% else %},{% endif %}
            {%- endfor %}
        )

    {% when Type::Optional with (inner_type) -%}
    # The Optional<T> type for {{ inner_type.canonical_name() }}.

    @classmethod
    def read{{ canonical_type_name }}(cls, stream):
        flag = stream._unpack_from(1, ">b")
        if flag == 0:
            return None
        elif flag == 1:
            return cls.read{{ inner_type.canonical_name()|class_name_py }}(stream)
        else:
            raise InternalError("Unexpected flag byte for {{ canonical_type_name }}")

    {% when Type::Sequence with (inner_type) -%}
    # The Sequence<T> type for {{ inner_type.canonical_name() }}.

    @staticmethod
    def read{{ canonical_type_name }}(stream):
        count = stream._unpack_from(4, ">i")
        if count < 0:
            raise InternalError("Unexpected negative sequence length")
        items = []
        while count > 0:
            items.append(stream.read{{ inner_type.canonical_name()|class_name_py }}())
            count -= 1
        return items

    {% when Type::Map with (inner_type) -%}
    # The Map<T> type for {{ inner_type.canonical_name() }}.

    @classmethod
    def read{{ canonical_type_name }}(cls, stream):
        count = stream._unpack_from(4, ">i")
        if count < 0:
            raise InternalError("Unexpected negative map size")
        items = {}
        while count > 0:
            key = cls.readString(stream)
            items[key] = stream.read{{ inner_type.canonical_name()|class_name_py }}()
            count -= 1
        return items

    {%- else -%}
    # This type cannot currently be serialized, but we can produce a helpful error.
    @staticmethod
    def read{{ canonical_type_name }}(stream):
        raise InternalError("RustBufferStream.read not implemented yet for {{ canonical_type_name }}")

    {%- endmatch -%}
    {%- endfor %}

File: uniffi_bindgen/src/bindings/ruby/templates/RustBufferStream.rb
# Mixin containing type-specific read methods as module functions.
# Call sites name this module explicitly (e.g.
# `::ThisNs::RustBufferStreamMixin.read_TypeFoo(stream)`) so two crates
# that both define `Foo` do not flatten `read_TypeFoo` onto one receiver.
# The `stream` argument is always the *local* crate's RustBufferStream so
# unpack_from / read route through that crate's buffer.
#
# Nested types from further crates are reached by lexical calls in this
# mixin's generated bodies, not by `include`. Do not use `module_function`.
# InternalError in these methods is this crate's class.
module RustBufferStreamMixin
  {% for typ in ci.iter_local_types() -%}
  {%- let canonical_type_name = self::canonical_name(typ) -%}
  {%- match typ -%}

  {% when Type::Int8 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 1, 'c'
  end

  {% when Type::UInt8 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 1, 'c'
  end

  {% when Type::Int16 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 2, 's>'
  end

  {% when Type::UInt16 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 2, 'S>'
  end

  {% when Type::Int32 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 4, 'l>'
  end

  {% when Type::UInt32 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 4, 'L>'
  end

  {% when Type::Int64 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 8, 'q>'
  end

  {% when Type::UInt64 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 8, 'Q>'
  end

  {% when Type::Float32 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 4, 'g'
  end

  {% when Type::Float64 -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    stream.unpack_from 8, 'G'
  end

  {% when Type::Boolean -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    v = stream.unpack_from 1, 'c'

    return false if v == 0
    return true if v == 1

    raise InternalError, 'Unexpected byte for Boolean type'
  end

  {% when Type::String -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    size = stream.unpack_from 4, 'l>'

    raise InternalError, 'Unexpected negative string length' if size.negative?

    stream.read(size).force_encoding(Encoding::UTF_8)
  end

  {% when Type::Bytes -%}

  def self.read_{{ self::canonical_name(typ) }}(stream)
    size = stream.unpack_from 4, 'l>'

    raise InternalError, 'Unexpected negative byte string length' if size.negative?

    stream.read(size).force_encoding(Encoding::BINARY)
  end

  {% when Type::Timestamp -%}
  # The Timestamp type.
  ONE_SECOND_IN_NANOSECONDS = 10**9

  def self.read_{{ canonical_type_name }}(stream)
    seconds = stream.unpack_from 8, 'q>'
    nanoseconds = stream.unpack_from 4, 'L>'

    # UniFFi conventions assume that nanoseconds part has to represent nanoseconds portion of
    # duration between epoch and the timestamp moment. Ruby `Time#tv_nsec` returns the number of
    # nanoseconds for the subsecond part, which is sort of opposite to "duration" meaning.
    # Hence we need to convert value returned by `Time#tv_nsec` back and forth with the following
    # logic:
    if seconds < 0 && nanoseconds != 0
      # In order to get duration nsec we shift by 1 second:
      nanoseconds = ONE_SECOND_IN_NANOSECONDS - nanoseconds

      # Then we compensate 1 second shift:
      seconds -= 1
    end

    Time.at(seconds, nanoseconds, :nanosecond, in: '+00:00').utc
  end

  {% when Type::Duration -%}
  # The Duration type.

  def self.read_{{ canonical_type_name }}(stream)
    seconds = stream.unpack_from 8, 'q>'
    nanoseconds = stream.unpack_from 4, 'L>'

    Time.at(seconds, nanoseconds, :nanosecond, in: '+00:00').utc
  end

  {% when Type::Object with { name: object_name, .. } -%}
  # The Object type {{ object_name }}.

  def self.read_{{ canonical_type_name }}(stream)
    handle = stream.unpack_from 8, 'Q>'
    return {{ object_name|class_name_rb }}.uniffi_lift(handle)
  end

  {% when Type::Enum { name, .. } -%}
  {%- let e = ci.get_enum_definition(name).unwrap() -%}
  {% if !ci.is_name_used_as_error(name) %}
  {% let enum_name = name %}
  # The Enum type {{ enum_name }}.

  def self.read_{{ canonical_type_name }}(stream)
    variant = stream.unpack_from 4, 'l>'
    {% if e.is_flat() -%}
    {%- for variant in e.variants() %}
    if variant == {{ loop.index }}
      return {{ enum_name|class_name_rb }}::{{ variant.name()|enum_name_rb }}
    end
    {%- endfor %}

    raise InternalError, 'Unexpected variant tag for {{ canonical_type_name }}'
    {%- else -%}
    {%- for variant in e.variants() %}
    if variant == {{ loop.index }}
        {%- if variant.has_fields() %}
        {%- let named_fields = !variant.fields()[0].name().is_empty() %}
        return {{ enum_name|class_name_rb }}::{{ variant.name()|enum_name_rb }}.new(
            {%- for field in variant.fields() %}
            {% if named_fields %}{{ field.name()|var_name_rb }}: {% endif %}{{ self.rust_buffer_read(field.as_type().borrow())? }}(stream){% if loop.last %}{% else %},{% endif %}
            {%- endfor %}
        )
        {%- else %}
        return {{ enum_name|class_name_rb }}::{{ variant.name()|enum_name_rb }}.new
        {% endif %}
    end
    {%- endfor %}
    raise InternalError, 'Unexpected variant tag for {{ canonical_type_name }}'
    {%- endif %}
  end

  {% else %}

  {% let error_name = name %}

  # The Error type {{ error_name }}

  def self.read_{{ canonical_type_name }}(stream)
    variant = stream.unpack_from 4, 'l>'
    {% if e.is_flat() -%}
    {%- for variant in e.variants() %}
    if variant == {{ loop.index }}
      return {{ error_name|class_name_rb }}::{{ variant.name()|class_name_rb }}.new(
        {{ self.rust_buffer_read(&Type::String)? }}(stream)
      )
    end
    {%- endfor %}

    raise InternalError, 'Unexpected variant tag for {{ canonical_type_name }}'
    {%- else -%}
    {%- for variant in e.variants() %}
    if variant == {{ loop.index }}
        {%- if variant.has_fields() %}
        {%- let named_fields = !variant.fields()[0].name().is_empty() %}
        return {{ error_name|class_name_rb }}::{{ variant.name()|class_name_rb }}.new(
            {%- for field in variant.fields() %}
            {% if named_fields %}{{ field.name()|var_name_rb }}: {% endif %}{{ self.rust_buffer_read(field.as_type().borrow())? }}(stream){% if loop.last %}{% else %},{% endif %}
            {%- endfor %}
        )
        {%- else %}
        return {{ error_name|class_name_rb }}::{{ variant.name()|class_name_rb }}.new
        {%- endif %}
    end
    {%- endfor %}

    raise InternalError, 'Unexpected variant tag for {{ canonical_type_name }}'
    {%- endif %}
  end
  {% endif %}

  {% when Type::Record { name: record_name, .. } -%}
  {%- let rec = ci.get_record_definition(record_name).unwrap() -%}
  # The Record type {{ record_name }}.

  def self.read_{{ canonical_type_name }}(stream)
    {{ rec.name()|class_name_rb }}.new(
      {%- for field in rec.fields() %}
      {{ field.name()|var_name_rb }}: {{ self.rust_buffer_read(field.as_type().borrow())? }}(stream){% if loop.last %}{% else %},{% endif %}
      {%- endfor %}
    )
  end

  {% when Type::Optional { inner_type } %}
  # The Optional<T> type for {{ self::canonical_name(inner_type) }}.

  def self.read_{{ canonical_type_name }}(stream)
    flag = stream.unpack_from 1, 'c'

    if flag == 0
      return nil
    elsif flag == 1
      return {{ self.rust_buffer_read(inner_type)? }}(stream)
    else
      raise InternalError, 'Unexpected flag byte for {{ canonical_type_name }}'
    end
  end

  {% when Type::Sequence { inner_type } -%}
  # The Sequence<T> type for {{ self::canonical_name(inner_type) }}.

  def self.read_{{ canonical_type_name }}(stream)
    count = stream.unpack_from 4, 'l>'

    raise InternalError, 'Unexpected negative sequence length' if count.negative?

    items = []

    count.times do
      items.append {{ self.rust_buffer_read(inner_type)? }}(stream)
    end

    items
  end

  {% when Type::Set { inner_type } -%}
  # The Set<T> type for {{ self::canonical_name(inner_type) }}.

  def self.read_{{ canonical_type_name }}(stream)
    count = stream.unpack_from 4, 'l>'

    raise InternalError, 'Unexpected negative set size' if count.negative?

    items = Set.new

    count.times do
      items.add {{ self.rust_buffer_read(inner_type)? }}(stream)
    end

    items
  end

  {% when Type::Map { key_type: k, value_type: v } -%}
  # The Map<T> type for {{ canonical_type_name }}.

  def self.read_{{ canonical_type_name }}(stream)
    count = stream.unpack_from 4, 'l>'
    raise InternalError, 'Unexpected negative map size' if count.negative?

    items = {}
    count.times do
      key = {{ self.rust_buffer_read(k)? }}(stream)
      items[key] = {{ self.rust_buffer_read(v)? }}(stream)
    end

    items
  end

  {% when Type::Custom { name, builtin, .. } -%}
  {%- match config.custom_types.get(name.as_str()) %}
  {%- when Some(cfg) %}{%- if cfg.has_conversion() %}
  # Custom type {{ name }}: reads builtin `{{ self::canonical_name(builtin) }}`, then applies lift.
  def self.read_{{ canonical_type_name }}(stream)
    raw = {{ self.rust_buffer_read(builtin)? }}(stream)
    {{ cfg.lift("raw") }}
  end
  {%- else %}
  # The Custom type {{ name }} delegates deserialization to its builtin type.
  def self.read_{{ canonical_type_name }}(stream)
    {{ self.rust_buffer_read(builtin)? }}(stream)
  end
  {%- endif %}
  {%- when None %}
  # The Custom type {{ name }} delegates deserialization to its builtin type.
  def self.read_{{ canonical_type_name }}(stream)
    {{ self.rust_buffer_read(builtin)? }}(stream)
  end
  {%- endmatch %}

  {% when Type::CallbackInterface { name, .. } -%}

  # The CallbackInterface type {{ name }}: read a uint64 handle.
  def self.read_{{ canonical_type_name }}(stream)
    handle = stream.unpack_from 8, 'Q>'
    {{ self::canonical_name(typ) }}FfiConverter.lift handle
  end

  {%- else -%}
  # This type is not yet supported in the Ruby backend.
  def self.read_{{ canonical_type_name }}(stream)
    raise InternalError, 'RustBufferStream.read not implemented yet for {{ canonical_type_name }}'
  end

  {%- endmatch -%}
  {%- endfor %}
end

# Helper for structured reading of values from a RustBuffer.
class RustBufferStream
  def initialize(rbuf)
    @rbuf = rbuf
    @offset = 0
  end

  def remaining
    @rbuf.len - @offset
  end

  def read(size)
    raise InternalError, 'read past end of rust buffer' if @offset + size > @rbuf.len

    data = @rbuf.data.get_bytes @offset, size

    @offset += size

    data
  end

  def unpack_from(size, format)
    raise InternalError, 'read past end of rust buffer' if @offset + size > @rbuf.len

    value = @rbuf.data.get_bytes(@offset, size).unpack format

    @offset += size

    # TODO: verify this
    raise 'more than one element!!!' if value.size > 1

    value[0]
  end
end

private_constant :RustBufferStream
